File: algorithms/merge_sort/912_sort_an_array/solution.py
from typing import List
import numpy as np


def merge(arr: List[int], left: int, middle: int, right: int):
    n_left = middle - left + 1
    n_right = right - middle

    subarr_left = []
    subarr_right = []

    for i in range(n_left):
        subarr_left.append(arr[left + i])

    for j in range(n_right):
        subarr_right.append(arr[middle + j + 1])

    subarr_left.append(np.infty)
    subarr_right.append(np.infty)

    i = 0
    j = 0
    k = left

    while k <= right:
        if subarr_left[i] <= subarr_right[j]:
            arr[k] = subarr_left[i]
            i += 1
        else:
            arr[k] = subarr_right[j]
            j += 1

        k += 1


def merge_sort(arr: List[int], left: int, right: int):
    if left < right:
        middle = (left + right) // 2
        merge_sort(arr, left, middle)
        merge_sort(arr, middle + 1, right)
        merge(arr, left, middle, right)


def sort_array(arr: List[int]) -> List[int]:
    merge_sort(arr, 0, len(arr) - 1)
    return arr


n = input()
nums = []

for _ in range(int(n)):
    nums.append(int(input()))

nums_new = sort_array(nums)

print(*nums_new)

File: algorithms/merge_sort/912_sort_an_array/solution.cpp
#include <iostream>
#include <vector>
#include <limits>

void merge(std::vector<int>& A, const int left, const int middle, const int right)
{
    const int n_1 = middle - left + 1;
    const int n_2 = right - middle;

    std::vector<int> L(n_1 + 1), R(n_2 + 1);

    for(int i = 0; i < n_1; i++)
        L[i] = A[left + i];

    for(int j = 0; j < n_2; j++)
        R[j] = A[middle + j + 1];

    L[n_1] = std::numeric_limits<int>::max();
    R[n_2] = std::numeric_limits<int>::max();

    int i = 0;
    int j = 0;

    for(int k = left; k <= right; k++)
    {
        if(L[i] <= R[j])
        {
            A[k] = L[i];
            i++;
        } else
        {
            A[k] = R[j];
            j++;
        }
    }
}

void merge_sort(std::vector<int>& A, const int left, const int right)
{
    if(left < right)
    {
        int middle = (left + right) / 2;
        merge_sort(A, left, middle);
        merge_sort(A, middle + 1, right);
        merge(A, left, middle, right);
    }
}

std::vector<int> sort_array(std::vector<int>& nums) 
{
    merge_sort(nums, 0, nums.size() - 1);
    return nums;
}

int main()
{
    int n {};
    std::cin >> n;

    std::vector<int> v(n);
    for(auto& i : v)
        std::cin >> i;

    v_new = sort_array(v);
    for(const auto& i : v_new)
        std::cout << i << " ";

    return 0;
}
